torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -1961,6 +1961,41 @@ std::tuple<Tensor, Tensor> atan2_backward(const Tensor& grad, const Tensor& self
             output_mask[1] ? grad * -self * recip : Tensor() };
 }
 
+std::tuple<Tensor, Tensor> attn_backward(const Tensor& grad_output, const Tensor& grad_attn,
+                                         const Tensor& q, const Tensor& k, const Tensor& v,
+                                         const Tensor& attn) {
+  bool grad_output_defined = grad_output.defined();
+  bool grad_attn_defined = grad_attn.defined();
+
+  bool in_place;
+  Tensor grad_u3;
+  if (grad_output_defined) {
+    grad_u3 = grad_output.matmul(v.t());
+    in_place = true;
+    if (grad_attn_defined) {
+      grad_u3.add_(grad_attn);
+    }
+  } else if (grad_attn_defined) {
+    grad_u3 = grad_attn;
+    in_place = false;
+  } else {
+    return std::make_tuple(Tensor(), Tensor());
+  }
+
+  TORCH_INTERNAL_ASSERT(grad_u3.defined());
+
+  // Conditionally point-wise multiplies tensors in-place.
+  auto mul = [](bool in_place, Tensor a, const Tensor& b) {
```

Comment:
I guess this is not needed anymore as you use it only once?

Reply:
Done. FWIW, I think it reads better with the function, but it's subjective and marginal either way.